### PR TITLE
Multiplicative Integration Recurrent Neural Networks

### PR DESCRIPTION
**Pull request by cydonia999**
This is the same PR as #9286(closed) after I
   - resolved conflicts
   - modified __init__.py
   - solved cla:no problem(I hope)

I implemented Multiplicative Integration variants of recurrent neural networks
(RNN, GRU and LSTM) proposed in

Yuhuai Wu, Saizheng Zhang, Ying Zhang, Yoshua Bengio, Ruslan Salakhutdinov,
On Multiplicative Integration with Recurrent Neural Networks. NIPS, 2016.
https://arxiv.org/abs/1606.06630

The RNNs proposed in the paper are implemented as:

    MultiplicativeIntegrationRNNCell
    MultiplicativeIntegrationGRUCell
    MultiplicativeIntegrationLSTMCell
    _multiplicative_integration as a helper function
in
    tensorflow/contrib/rnn/python/ops/rnn_cell.py
    tensorflow/contrib/rnn/__init__.py

Test codes:
    tensorflow/contrib/rnn/python/kernel_tests/rnn_cell_test.py

### CODE REVIEW

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@ebrevdo what do you think?

**Comment by ebrevdo** [comment]
Needs some cleanup but seems ok to accept eventually.

Please remove the input_size arguments and change __call__ to call, and
rename "weights" to kernel and "biases" to bias.   See other rnncells in
the same file for more examples on style.

File: README.md [hunk 1]
```diff
@@ -1,3 +1,25 @@
+# Multiplicative Integration Recurrent Neural Networks in Tensorflow
```
Comment (ebrevdo) [inline README.md]: please revert the change to this file.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 2]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+
+  The implementation is based on:
+    https://arxiv.org/abs/1606.06630
+
+    Yuhuai Wu, Saizheng Zhang, Ying Zhang, Yoshua Bengio, Ruslan Salakhutdinov,
+    On Multiplicative Integration with Recurrent Neural Networks. NIPS, 2016.
+  """
+
+  def __init__(self, num_units, bias_start=0.0, alpha_start=1.0, 
+                beta_start=1.0, activation=math_ops.tanh, reuse=None):
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit: whitespace

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 3]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+
+  The implementation is based on:
+    https://arxiv.org/abs/1606.06630
+
+    Yuhuai Wu, Saizheng Zhang, Ying Zhang, Yoshua Bengio, Ruslan Salakhutdinov,
+    On Multiplicative Integration with Recurrent Neural Networks. NIPS, 2016.
+  """
+
+  def __init__(self, num_units, bias_start=0.0, alpha_start=1.0, beta_start=1.0,
+      activation=math_ops.tanh, reuse=None):
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit: whitespace

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 4]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+  @property
+  def output_size(self):
+    return self._num_units
+
+  def call(self, inputs, state):
+    """Run one step of Multiplicative Integration Gated Recurrent Unit(GRU)."""
+    with vs.variable_scope("gates"):
+      r, u = array_ops.split(
+          value=_multiplicative_integration(
+            [inputs, state], [self._num_units, self._num_units],
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit; whitepace

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 5]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+
+  The implementation is based on:
+    https://arxiv.org/abs/1606.06630
+
+    Yuhuai Wu, Saizheng Zhang, Ying Zhang, Yoshua Bengio, Ruslan Salakhutdinov,
+    On Multiplicative Integration with Recurrent Neural Networks. NIPS, 2016.
+  """
+
+  def __init__(self, num_units, forget_bias=0.0, bias_start=0.0, alpha_start=1.0, 
+      beta_start=1.0, state_is_tuple=True, activation=math_ops.tanh, reuse=None):
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit: whitespace

Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: remove state_is_tuple argument; always assume it's True.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 6]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+  @property
+  def output_size(self):
+    return self._num_units
+
+  def call(self, inputs, state):
+    """Run one step of Multiplicative Integration Long short-term memory(LSTM)."""
+    if self._state_is_tuple:
+      c, h = state
+    else:
+      c, h = array_ops.split(value=state, num_or_size_splits=2, axis=1)
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: remove this code path.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 7]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+
+    if self._state_is_tuple:
+      new_state = rnn_cell_impl.LSTMStateTuple(new_c, new_h)
+    else:
+      new_state = array_ops.concat([new_c, new_h], 1)
+    return new_h, new_state
+
+
+def _multiplicative_integration(args, output_sizes, bias, bias_start=0.0, 
+    alpha_start=1.0, beta_start=1.0):
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit: whitespace

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 8]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+  Raises:
+    ValueError: if some of the arguments has unspecified or wrong shape.
+  """
+  if args is None or (nest.is_sequence(args) and not args):
+    raise ValueError("`args` must be specified")
+  if not nest.is_sequence(args):
+    raise ValueError("`args` must be list")
+  if len(args) != 2:
+    raise ValueError("`args` must contain 2 tensors")
+  if output_sizes is None or (nest.is_sequence(output_sizes)
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: `if not output_sizes`?

the only case you're allowing here is output_sizes == 0; and do you want to allow that?

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 9]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+    w1, w2 = array_ops.split(kernel, num_or_size_splits=arg_sizes, axis=0)
+    b1, b2 = array_ops.split(betas, num_or_size_splits=2, axis=0)
+    wx1, wx2 = math_ops.matmul(args[0], w1), math_ops.matmul(args[1], w2)
+    res = alphas * wx1 * wx2 + b1 * wx1 + b2 * wx2
+
+    if not bias:
+      return res
+    with vs.variable_scope(outer_scope) as inner_scope:
+      inner_scope.set_partitioner(None)
+      bias = vs.get_variable( "bias", [total_output_size], dtype=dtype,
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit: whitespace between `(` and `"bias"`

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 10]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+      raise ValueError("multiplicative_ntegration expects shape[1] to be "
+             "provided for shape %s, but saw %s" % (shape, shape[1]))
+    else:
+      total_arg_size += shape[1].value
+      arg_sizes.append(shape[1].value)
+
+  dtype = [a.dtype for a in args][0]
+
+  scope = vs.get_variable_scope()
+  with vs.variable_scope(scope) as outer_scope:
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: you don't need this "with" statement because you're not modifying outer_scope anywhere.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 11]
```diff
@@ -2095,3 +2095,293 @@ def call(self, inputs, state):
...
+    kernel = vs.get_variable( "kernel", [total_arg_size, total_output_size], 
+        dtype=dtype) 
+    alphas = vs.get_variable( "alphas", [total_output_size], dtype=dtype,
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: nit, here, above & below: space between `(` and `"`

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 8 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: No.
When output_sizes == 0, the next if clause raise an error:
>   if not nest.is_sequence(output_sizes):
>     raise ValueError("`output_sizes` must be list")

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 10 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: **outer_scope** is used for the bias:
```
  scope = vs.get_variable_scope()
  with vs.variable_scope(scope) as outer_scope:

(omit)

    with vs.variable_scope(outer_scope) as inner_scope:
      inner_scope.set_partitioner(None)
      bias = vs.get_variable("bias", [total_output_size], dtype=dtype,
          initializer=init_ops.constant_initializer(bias_start, dtype=dtype))
  return nn_ops.bias_add(res, bias)
```

Should I delete 
```
  scope = vs.get_variable_scope()
  with vs.variable_scope(scope) as outer_scope:
```
and
`    with vs.variable_scope(outer_scope) as inner_scope:` ?

**Comment by vrv** [comment]
ping for @cydonia999 (it looks like not all review comments have been addressed? please use the review tool to respond as needed).

File: README.md (hunk 1 above)
Comment (cydonia999) [inline README.md]: reverted.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 2 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Indentation adjusted.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 3 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Indentation adjusted.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 4 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Indentation adjusted.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 5 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Indentation adjusted.

Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: removed the state_is_tuple argument.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 6 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: removed these four lines, keeping `c, h = state`

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 7 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Indentation adjusted.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 9 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: removed the unnecessary whitespace.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 11 above)
Comment (cydonia999) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: removed the unnecessary spaces.

**Comment by cydonia999** [comment]
@vrv Thank you for alerting me to the review comments. I have replied all the review comments.

**Comment by rmlarsen** [comment]
@ebrevdo can you take another look when you are back in the office, please?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
I thinks this issue is waiting for @ebrevdo 's review.

**Comment by sb2nov** [comment]
@ebrevdo ping

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 12]
```diff
@@ -2095,3 +2095,263 @@ def call(self, inputs, state):
 
     new_state = rnn_cell_impl.LSTMStateTuple(c, m)
     return m, new_state
+
+
+class MultiplicativeIntegrationRNNCell(rnn_cell_impl.RNNCell):
```
Comment (ebrevdo) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Can you make this subclass LayerRNNCell, and move all the get_variable() calls into a "def build(input_shape):".  at the end of the build function set `self.built = True`.  In `call`, only access variables via `self` (i.e., `self._kernel`, `self._bias`, etc).  See the current LSTMCell and GRUCell implementations to see what I mean.

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 10 above)
Comment (ebrevdo) [reply tensorflow/contrib/rnn/python/ops/rnn_cell.py]: the inner_scope seems reasonable; since in it you're disabling the partitioner.  but you don't need the outer_scope because you're not modifying it at all.  so basically just do:

```
with vs.variable_scope(vs.get_variable_scope()) as inner_scope:
   inner_scope.set_partitioner(None)
   ... # set up bias
```

as above, move all this `get_variable` stuff into the `build` methods.

ping me via email (ebrevo@) when ready for re-review.

**Comment by drpngx** [comment]
@cydonia999 could you reply to @ebrevdo comments? Also, please rebase and push again.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by raghuraman-k** [comment]
@cydonia999 : Are you planning to work on this PR?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 44 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
